src/coprocessor/codec/batch/lazy_column.rs

```diff
@@ -143,47 +144,59 @@ impl LazyBatchColumn {
         }
     }
 
-    /// Retains the elements according to a boolean array.
-    ///
-    /// # Panics
-    ///
-    /// Panics if `retain_arr` is not long enough.
-    #[inline]
-    pub fn retain_by_array(&mut self, retain_arr: &[bool]) {
-        match self {
-            LazyBatchColumn::Raw(v) => v.retain_by_array(retain_arr),
-            LazyBatchColumn::Decoded(v) => v.retain_by_array(retain_arr),
-        }
-    }
-
-    /// Decodes this column in place if the column is not decoded.
+    /// Decodes this column if the column is not decoded, according to the given logical rows map.
+    /// After decoding, the decoded column will have the same physical layout as the encoded one
+    /// (i.e. the same logical rows), but elements in unnecessary positions will not be decoded
+    /// and will be `None`.
     ///
     /// The field type is needed because we use the same `DateTime` structure when handling
     /// Date, Time or Timestamp.
     // TODO: Maybe it's a better idea to assign different eval types for different date types.
-    pub fn ensure_decoded(&mut self, time_zone: &Tz, field_type: &FieldType) -> Result<()> {
+    pub fn ensure_decoded(
+        &mut self,
+        time_zone: &Tz,
+        field_type: &FieldType,
+        logical_rows: &[usize],
+    ) -> Result<()> {
         if self.is_decoded() {
             return Ok(());
         }
 
         let eval_type = box_try!(EvalType::try_from(field_type.tp()));
-
-        let mut decoded_column = VectorValue::with_capacity(self.capacity(), eval_type);
-        {
-            for value in self.raw().iter() {
-                decoded_column.push_datum(value, time_zone, field_type)?;
+        let raw_vec = self.raw();
+        let raw_vec_len = raw_vec.len();
+
+        let mut decoded_column = VectorValue::with_capacity(raw_vec.capacity(), eval_type);
```

Comment:
Is it OK to set the capacity to `raw_vec_len`? The raw vector can be truncated before decoded, then we waste the extra capacity. I cannot think of a case when we append to a decoded column converted from raw.

Reply:
Good catch. Previously we want to preserve the capacity after clone (and decode), but looks like in late materialization we no longer need this.